### DISCUSSION TITLE
Nested consult-outline?

### DISCUSSION THREAD

**Discussion by mekeor** [minad/consult#862] [Ideas]
I love using [`outline-minor-mode`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Regexp-Backslash.html), especially in personal projects, where I define `outline-regexp` and `outline-heading-alist` as directory-local variables, which allows for structuring files into hierarchical sections and navigating them via `consult-outline`. That Consult command features narrowing down the candidates to a level. I suggest to add two more features here:

1. Show the candidates as nesting paths, that is breadcrumbs, just like `consult-org-heading`. I'd suggest to do so when a univeral prefix argument (`C-u`) is given.
2. Allow the regular expressions in `outline-heading-alist` to extract the relevant part of the heading by using an explicitely numbered group with a certain number (see [`(info "(elisp) Regexp Backslash")`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Regexp-Backslash.html)). E.g. if your heading regular expression matches `;;; Title`, then you could extract the `Title` with a `;;; \(?1:.*\)$` or so.

What do you think?

**Comment by minad** [comment]
Hi, thanks for the proposals. Regarding your second proposal - are groups commonly used and documented for `outline-regexp`? If not, we should not add this here. I think the first proposal somehow relies on the second one, since otherwise we get too much noise when displaying the nested header?

**Reply by mekeor** [reply]
> are groups commonly used and documented for outline-regexp?

I don't think so.

> I think the first proposal somehow relies on the second one, since otherwise we get too much noise when displaying the nested header?

Yes, I agree.

**Reply by mekeor** [reply]
Should we close this discussion? I guess, the feature should be implemented in a separate package, given the circumstances?

**Reply by IceAsteroid** [reply]
Can ```consult-org-heading``` be achieved to not show prefixed parents of a sub-level candidate instead, but show asterisks to indicate levels, and if the search matches its parent, its sub-level headings also shows below it

In case if "consult-outline" doesn't do this

**Comment by okamsn** [comment]
You might be interested in this command from the Selectrum wiki:

https://github.com/radian-software/selectrum/wiki/Useful-Commands#jumping-to-outline-headings

I modified it for Consult, but I don't think that I added it to the Consult wiki. I will do that in a while.

**Reply by okamsn** [reply]
I've added it here: https://github.com/minad/consult/wiki#completing-a-hierarchical-outline

**Reply by IceAsteroid** [reply]
Could you please explain a bit about how it works?

I used in an org buffer, but consult-outline didn't seemingly change its behavior.

Here's a defined command "my-consult-outline-path" of the provided code, but it worked not as expected.

Here's the issue page that I posted for a similar question: #987

**Reply by okamsn** [reply]
The command `my-consult-outline-path` searches the buffer for lines matching the regexp for the major mode specified in `my-consult-outline-path-formats`. The first group of the regexp is a prefix string whose length determines the level of the line in the outline (such as sublevel, sub-sublevel, etc.). The second group in the regexp is  the section title. Based on the length of the prefix string, it concatenates section titles of the current and containing levels to make a completion candidate.

The command does not change the behavior of `consult-outline`, but it does fall back to it if there is no specifed regexp. For me, that has been enough.

Does that help?

**Reply by IceAsteroid** [reply]
@minad 
@mekeor If you have read my issue post #987 , you would know what I meant.

I've tried "my-consult-outline-path", but in org-mode, it still chunked that sub-level candidates are prefixed by its parents.

It's not instead like I described in my post, where:

For example:
```
* Colors
** Red
*** Apple
** Yellow
*** Banana
* Fruits
** Orange
** Apple
```

Say if I search for ```Colors```, the minibufer's candidates would be:

```
* Colors
** Red
*** Apple
** Yellow
*** Banana
```

If heading level is triggerd, say Level 2, the candidates would be:

```
* Colors
** Red
** Yellow
```

The reason why not to use ```consult-org-heading``` instead, is because it displays subheadings preceded by its parent headings.

When a subheading is deeply nested, the heading can't be displayed correctly as it's truncated at the edge, except its parent headings are shown.

It would however be(not desired):
```
first-level heading A/subheading1/subheading1.1/subheading1.1.1/sub->
first-level heading B/subheading1/subheading1.1/subheading1.1.1/sub->
```

In conclusion, if searching a heading would also produce its sub-headings as candidates below it, it'd still be clear of the levels of each heading, with even only asterisks in org, or semi-colons in elisp, etc,

**Reply by IceAsteroid** [reply]
@minad

How hard is it to implement such a feature? What I need if to do it?

I really want it though, the last puzzle of my org-mode setup

**Reply by hmelman** [reply]
Not sure I completely follow, but if you want the path merely for breadcrumb-like behavior perhaps the thing to do is use annotations.  Off the top of my head, define a new marginalia category "heading" and show the parent path as an annotation.  That way you'll still see the terminal heading untruncated and your annotation can display breadcrumb however (shortening or reordering if it's too long, possibly like `uniquify-buffer-name`).

**Reply by minad** [reply]
@IceAsteroid 

> In conclusion, if to search a heading would also produce its sub-headings as candidates below it, it'd still be clear of the levels of each heading, with even only asterisks in org, or semi-colons in elisp, etc,

Quite difficult. There are multiple ways to do that, depending on the exact UI you want to achieve. In any case, it is unlikely to fit nicely into the existing infrastructure or it would require a rewrite of consult-org*. As an alternative, you could maybe give org-ql a try.

**Reply by IceAsteroid** [reply]
Good news, I found the solution with less hassle to the feature I described.

Just for the solely need of searching in org-mode, I'm gonna refactor the 'consult-org-heading' instead of 'consult-outline'.

* First, make candidates prefixed with their level indicators like in 'consult-outline'.

* Second, change sequence of the display of ancestor headings of a candidate like follows.

```
* Fruits
** Red Color
*** Apple
*** Cherry
** Yellow Color
*** Banana
*** Lemon
* Drinks...
* Clothes...
....
```

To be displayed when searching `fruits` in the minibuffer
```
* Fruits
** Red Color/Fruits
*** Apple/Red Color/Fruits
*** Cherry/Red Color/Fruits
** Yellow Color/Fruits
*** Banana/Yellow Color/Fruits
*** Lemon/Yellow Color/Fruits
```

I'm gonna try to implement it, and post it here

Thank you for the package, and have good days.

**Reply by minad** [reply]
@IceAsteroid So what's the difference? The star prefixing and the reverse order of the headings? I think we could add this to Consult without too much effort - the stars by default and the reverse order guarded behind a flag. See https://github.com/minad/consult?tab=readme-ov-file#contributions regarding FSF assignment.

**Reply by IceAsteroid** [reply]
Sorry for the disturbance, sir. You may forget the details, as I said before, when too many nested ancestor headings of a sub-level heading, the actual text of the heading is truncated by default as a candidate in the minibuffer.

Some of my org files are very large, and nested headings and long headings are very frequently used and necessary for a large org file.

I also don't want to have `truncate-lines` enabled in `vertico`, since some behaviors are weird with vertico or probably just consult-org-heading, when short candidates don't exceed the screen, they're still shown as occupying two lines.

That's the reason I've been trying to have such a feature.

I'm trying to do it, the previous proposal might not be perfect, I've got a better idea, so you don't have to do it now.

I hope you understand, it's nothing offensive.

Have a nice day!

**Reply by minad** [reply]
Yes, I've not forgotten the motivation. It sounds reasonable to add reversing the header as an option to consult-org-heading. Adding the stars is also not a bad idea since one can see the level right away.

**Reply by IceAsteroid** [reply]
Thank you for the understanding, should I open a standalone issue, or a discussion?

My skills on elisp are very limited, I'm now in the process of trying to understand consult's code.

I have two more requests:
* Inherit headings' properties including the colors, the stars(since they are changed by the use of `org-superstar`. If possible, the ancestors prefixed in a candidate should also have their own corresponding colors as in the org file.
* Inherit overlays that are added to the headings in the org file(if possible), only for the current heading, not also for the ancestors in a candidate to prevent visual pollution.

If you're willing to do it, please do. Otherwise I'll do it, but it takes time.

**Reply by IceAsteroid** [reply]
@minad, what do you think?

**Reply by minad** [reply]
Feel free to create a patch. Inheriting properties is automatic (if Org uses `buffer-substring` for extraction, similar to `consult-line`), but inheriting overlays is out of scope.

**Reply by IceAsteroid** [reply]
OK, glad to hear that. Thank you for the great work and patience as always. Take care :)
